bftengine/src/bcstatetransfer/RangeValidationTree.hpp

```diff
@@ -0,0 +1,268 @@
+  std::string getDirectParentValueStr(RVBId rvb_id) const;
+  // Return the min RVB ID in the tree. Return 0 if tree is empty.
+  RVBId getMinRvbId() const;
+  // Return the max RVB ID in the tree. Return 0 if tree is empty.
+  RVBId getMaxRvbId() const;
+  bool empty() const { return (id_to_node_.size() == 0) ? true : false; }
+  const std::string getRootCurrentValueStr() const { return root_ ? root_->current_value_.toString() : ""; }
+  void printToLog(bool only_node_id) const noexcept;
+  bool validate() const noexcept;
+  void clear() noexcept;
+
+ public:
+  struct NodeVal {
+    static NodeVal_t kNodeValueMax_;
+    static NodeVal_t kNodeValueModulo_;
+    static NodeVal_t calcMaxValue(size_t val_size);
+    static NodeVal_t calcModulo(size_t val_size);
+
+    NodeVal(const std::shared_ptr<char[]>&& val, size_t size);
+    NodeVal(const char* val_ptr, size_t size);
+    NodeVal(const NodeVal_t* val);
+    NodeVal(const NodeVal_t& val);
+    NodeVal(const NodeVal_t&& val);
+    NodeVal();
+
+    NodeVal& operator+=(const NodeVal& other);
+    NodeVal& operator-=(const NodeVal& other);
+    bool operator!=(const NodeVal& other);
+    bool operator==(const NodeVal& other);
+
+    const NodeVal_t& getMaxVal() const { return kNodeValueMax_; }
+    const NodeVal_t& getVal() const { return val_; }
+    std::string toString() const noexcept;
+    std::string getDecoded() const noexcept;
+    size_t getSize() const { return val_.MinEncodedSize(); }
+
+    static constexpr size_t kDigestContextOutputSize = BLOCK_DIGEST_SIZE;
+    static constexpr std::array<char, kDigestContextOutputSize> initialValueZeroData{};
+
+    NodeVal_t val_;
+  };
+
+  struct NodeInfo {
+    NodeInfo(uint64_t node_id)
+        : level(node_id >> kNIDBitsPerRVBIndex),
+          rvb_index(node_id & kRvbIndexMask),
+          id((static_cast<uint64_t>(level) << kNIDBitsPerRVBIndex) | rvb_index) {}
+    NodeInfo(uint8_t l, uint64_t index)
+        : level(l),
+          rvb_index(index & kRvbIndexMask),
+          id((static_cast<uint64_t>(level) << kNIDBitsPerRVBIndex) | rvb_index) {}
+    NodeInfo() = delete;
+
+    bool operator<(NodeInfo& other) const noexcept {
+      return ((level <= other.level) || (rvb_index < other.rvb_index)) ? true : false;
+    }
+    bool operator!=(NodeInfo& other) const noexcept {
+      return ((level != other.level) || (rvb_index != other.rvb_index)) ? true : false;
+    }
+    std::string toString() const noexcept;
+
+    static constexpr size_t kNIDBitsPerLevel = 8;
+    static constexpr size_t kNIDBitsPerRVBIndex = ((sizeof(uint64_t) * 8) - kNIDBitsPerLevel);
+    static constexpr size_t kMaxLevels = ((0x1 << kNIDBitsPerLevel) - 1);
+    static constexpr uint64_t kRvbIdMask = (kMaxLevels) << kNIDBitsPerRVBIndex;
+    static constexpr uint64_t kRvbIndexMask = std::numeric_limits<uint64_t>::max() & (~kRvbIdMask);
+
+    // node location in the tree
+    const uint64_t level : kNIDBitsPerLevel;
```

Comment:
Please add a TODO above the 3 fields.
The reason we had a bit field is to avoid what was written here.
level + rvb_index are id, and id is level + rvb_index.
And it is possible to cast  between them  with a helper function.
The 3 of them should catch 64 bit in space, and the level should be placed in the MSB part.
This should be added as a new task.

Many ways to implement the idea. for example:
union data {
 struct loc {
    const uint64_t rvb_index : kNIDBitsPerRVBIndex;
    const uint64_t level : kNIDBitsPerLevel;
 };
 const uint64_t id;
};